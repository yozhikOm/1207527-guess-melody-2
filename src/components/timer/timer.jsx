import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {createTimer} from '../../utils/createTimer.js';

class Timer extends PureComponent {
  constructor(props) {
    super(props);
  }

  /* _tickTimer() {
    const {gameTime, onTimerTick, onTimeExpired} = this.props;
    if (gameTime === 0) {
      clearInterval(this._gameTimer);
      return onTimeExpired();
    }
    return onTimerTick();
  }
  */

  _formatTime(timeInSeconds) {
    let minutes = Math.floor(timeInSeconds / 60) % 60;
    let seconds = Math.floor(timeInSeconds - minutes * 60);

    const time = {
      "minutes": minutes,
      "seconds": seconds
    };

    return time;
  }

  _onTimeExpired() {
    clearInterval(this._gameTimer);
    this.props.onTimeExpired();
  }

  componentDidMount() {
    /* this._gameTimer = setInterval(this._tickTimer.bind(this), 1000);*/

    const {gameTime, storeRemainingTime} = this.props;

    const timer = createTimer(gameTime, this._onTimeExpired.bind(this));

    this._gameTimer = setInterval(() => {
      storeRemainingTime(timer.tick());
    }, 1000);
  }

  componentWillUnmount() {
    this._onTimeExpired();
  }

  render() {
    const {gameTime} = this.props;
    const time = this._formatTime(gameTime);

    return (
      <div className="timer__value" xmlns="http://www.w3.org/1999/xhtml">
        <span className="timer__mins">{time.minutes >= 10 ? time.minutes : `0${time.minutes}`}</span>
        <span className="timer__dots">:</span>
        <span className="timer__secs">{time.seconds >= 10 ? time.seconds : `0${time.seconds}`}</span>
      </div>
    );
  }
}

Timer.propTypes = {
  gameTime: PropTypes.number.isRequired,
  storeRemainingTime: PropTypes.func.isRequired,
  onTimeExpired: PropTypes.func.isRequired,
};

export {Timer};
