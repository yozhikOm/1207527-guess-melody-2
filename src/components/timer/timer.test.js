import React from 'react';
import {Timer} from './timer.jsx';
import renderer from 'react-test-renderer';

it(`Timer компонент рендерится корректно`, () => {
  const timerComponent = renderer.create(
      <Timer
        gameTime={60}
        storeRemainingTime={jest.fn()}
        onTimeExpired={jest.fn()}
      />).toJSON();
  expect(timerComponent).toMatchSnapshot();
});
