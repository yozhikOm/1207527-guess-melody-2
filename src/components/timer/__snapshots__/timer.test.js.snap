// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer компонент рендерится корректно 1`] = `
<div
  className="timer__value"
  xmlns="http://www.w3.org/1999/xhtml"
>
  <span
    className="timer__mins"
  >
    01
  </span>
  <span
    className="timer__dots"
  >
    :
  </span>
  <span
    className="timer__secs"
  >
    00
  </span>
</div>
`;
